recipes/newrelic/apm/php-agent/php.yml

```diff
@@ -29,20 +29,80 @@ inputVars:
 install:
   version: "3"
 
+  vars:
+    TMP_INSTALL_DIR:
+      sh: mktemp -d /tmp/newrelic-php-agent.XXXXXX
+
   env:
     NEW_RELIC_DISTRIBUTED_TRACING_ENABLED: true
 
   tasks:
     default:
       cmds:
-        - task: install_packages
         - task: create_shared
+        - task: detect_services
+        - task: install_packages
         - task: config_newrelic_ini
         - task: finalize_agent_config
         - task: restart_services
         - task: send_transaction
         - task: delete_shared
 
+    create_shared:
+      cmds:
+        - rm -rf {{.TMP_INSTALL_DIR}}
+        - mkdir -p {{.TMP_INSTALL_DIR}}
+        - echo "Created temporary directory {{.TMP_INSTALL_DIR}} for installation."
+
+    detect_services:
+      cmds:
+        - echo "Detect services."
+        - |
+          cd {{.TMP_INSTALL_DIR}}
+          rm -f {{.TMP_INSTALL_DIR}}/nonpriv_users.txt
+          rm -f {{.TMP_INSTALL_DIR}}/processes_to_restart.txt
+
+          #
+          # For guided installs, we currently support `php-fpm`, `apache`, and `nginx`.
+          # There is a reason for the order.  According our install script:
+          # If both Apache and FPM are installed, we want FPM to win:
+          # while there are many ways to end up with libapache2-mod-php5 installed,
+          # it's unlikely php5-fpm will be installed unless FPM is actually in use.
```

Comment:
If both FPM and Apache are present, FPM is likely in use (or at least intended to be in use), but any case where it ends up not being used is a potential failed installed. In my testing of the deploy configs, I ran into several cases where I had PHP-FPM installed, but had not configured Apache properly to actually use it. I recommend that we either restart both or figure out whether Apache is configured to use FPM (e.g. parse configs, run a phpinfo to look for "Server API", or grep our logs for the sapi in use - this last one requires that we run a non-php-cli process).

Reply:
Currently the logic is detect fpm, apache, nginx in that order.
it will restart all that it detects.
The order really only matters in that only the first detected non-privileged user is used to run the php --ini at the end.